x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/MultiTypeEsField.java

```diff
@@ -0,0 +1,125 @@
+import org.elasticsearch.xpack.ql.type.DataTypes;
+import org.elasticsearch.xpack.ql.type.EsField;
+import org.elasticsearch.xpack.ql.type.InvalidMappedField;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * During IndexResolution it could occur that the same field is mapped to different types in different indices.
+ * The class MultiTypeEfField.UnresolvedField holds that information and allows for later resolution of the field
+ * to a single type during LogicalPlanOptimization.
+ * If the plan contains conversion expressions for the different types, the resolution will be done using the conversion expressions,
+ * in which case a MultiTypeEsField will be created to encapsulate the type resolution capabilities.
+ * This class can be communicated to the data nodes and used during physical planning to influence field extraction so that
+ * type conversion is done at the data node level.
+ */
+public class MultiTypeEsField extends EsField {
+    private final Map<String, Expression> indexToConversionExpressions;
+
+    public MultiTypeEsField(String name, DataType dataType, boolean aggregatable, Map<String, Expression> indexToConversionExpressions) {
+        super(name, dataType, Map.of(), aggregatable);
+        this.indexToConversionExpressions = indexToConversionExpressions;
+    }
+
+    public Map<String, Expression> getIndexToConversionExpressions() {
+        return indexToConversionExpressions;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (super.equals(obj) == false) {
+            return false;
+        }
+        if (obj instanceof MultiTypeEsField other) {
+            return super.equals(other) && indexToConversionExpressions.equals(other.indexToConversionExpressions);
+        }
+        return false;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), indexToConversionExpressions);
+    }
+
+    public Expression getConversionExpressionForIndex(String indexName) {
+        return indexToConversionExpressions.get(indexName);
+    }
+
+    /**
+     * During IndexResolution it could occur that the same field is mapped to different types in different indices.
+     * This class holds that information and allows for later resolution of the field to a single type during LogicalPlanOptimization.
+     * If the plan contains conversion expressions for the different types, the resolution will be done using the conversion expressions,
+     * in which case a MultiTypeEsField will be created to encapsulate the type resolution.
+     * If type resolution is not possible, due to the plan not containing explicit type conversion functions, then this class will
+     * be used to communicate the type resolution failure, since it's parent class is InvalidMappedField.
+     */
+    public static class UnresolvedField extends InvalidMappedField {
```

Comment:
Why not use InvalidMappedField directly since that is already picked up by the analyzer/verifier.

Reply:
This conversation has come full circle now, due to the recent reviews asking to not edit QL directly. After the above comment, I switched to editing QL directly. There were two classes where we edited QL `FieldAttribute` where the equals and hashcode were updated to include the underlying field, so query plan re-writing would work (could be seen as an oversight in QL code, so perhaps even a bug-fix), and the above mentioned InvalidMappedField. An attempt to completely extricate these two from QL failed (dependencies increased scope dramatically, turning a moderate PR into a monster). A second attempt at a compromise by making ESQL versions that maintained the simple name, and extended the QL versions worked for both classes, but was quite messy for FieldAttribute. After discussing with the team, we decided to keep the ESQL port of InvalidMappedField, but revert to the direct edit of the FieldAttribute. The thinking here is the up-coming split of QL and ESQL would be slightly facilitated by not having any edits in InvalidMappedField, while the edit to FieldAttribute we would probably want to keep anyway.